### PR TITLE
fix: drop unused modality columns in dataloader for cross-modal tasks

### LINKED ISSUE #4436: CIRRIT2IRetrieval errored out
### Describe the bug

```
Evaluating task CIRRIT2IRetrieval:   0%|          | 0/1 [02:27<?, ?it/s]
Traceback (most recent call last):
  File "/data/home/niklas/isaac/mteb/.venv/bin/mteb", line 10, in <module>
    sys.exit(main())
  File "/data/home/niklas/isaac/mteb/mteb/cli/build_cli.py", line 497, in main
    args.func(args)
  File "/data/home/niklas/isaac/mteb/mteb/cli/build_cli.py", line 91, in run
    mteb.evaluate(
  File "/data/home/niklas/isaac/mteb/mteb/evaluate.py", line 390, in evaluate
    _res = evaluate(
  File "/data/home/niklas/isaac/mteb/mteb/evaluate.py", line 489, in evaluate
    result = _evaluate_task(
  File "/data/home/niklas/isaac/mteb/mteb/evaluate.py", line 162, in _evaluate_task
    task_results[split] = task.evaluate(
  File "/data/home/niklas/isaac/mteb/mteb/abstasks/retrieval.py", line 335, in evaluate
    return super().evaluate(
  File "/data/home/niklas/isaac/mteb/mteb/abstasks/abstask.py", line 209, in evaluate
    scores[hf_subset] = self._evaluate_subset(
  File "/data/home/niklas/isaac/mteb/mteb/abstasks/retrieval.py", line 403, in _evaluate_subset
    results = retriever(
  File "/data/home/niklas/isaac/mteb/mteb/_evaluators/retrieval_evaluator.py", line 71, in __call__
    return search_model.search(
  File "/data/home/niklas/isaac/mteb/mteb/models/search_wrappers.py", line 183, in search
    result_heaps = self._full_corpus_search(
  File "/data/home/niklas/isaac/mteb/mteb/models/search_wrappers.py", line 254, in _full_corpus_search
    sub_corpus_embeddings = self.model.encode(
  File "/data/home/niklas/isaac/mteb/mteb/models/model_implementations/lco_embedding_models.py", line 125, in encode
    for batch in tqdm(inputs, disable=not show_progress_bar):
  File "/data/home/niklas/isaac/mteb/.venv/lib/python3.10/site-packages/tqdm/std.py", line 1181, in __iter__
    for obj in iterable:
  File "/data/home/niklas/isaac/mteb/.venv/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 741, in __next__
    data = self._next_data()
  File "/data/home/niklas/isaac/mteb/.venv/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 801, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/data/home/niklas/isaac/mteb/.venv/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 57, in fetch
    return self.collate_fn(data)
  File "/data/home/niklas/isaac/mteb/mteb/_create_dataloaders.py", line 235, in _custom_collate_fn
    raise ValueError(f"Found None in batch for key '{key}'")
ValueError: Found None in batch for key 'text'
```

### To reproduce

`mteb run -t CIRRIT2IRetrieval -m LCO-Embedding/LCO-Embedding-Omni-3B`

### Additional information

_No response_

### Are you interested to contribute a fix for this bug?

Yes

### PR DESCRIPTION
**Pull request by isaac-chung**
Fixes #4436

Cross-modal retrieval tasks (e.g. `CIRRIT2IRetrieval`, `Fashion200kI2TRetrieval`, `VisualNewsI2TRetrieval`) have `None` values  for modalities not used by that side of the retrieval — for example, `text=None` in an image-only corpus (`it2i`), or `image=None` in  text-only queries (`i2t`).

### Changes
- Drop columns for modalities not needed for the current prompt type in `_prepare_dataset`, using the task category (e.g. `it2i`) to determine which modalities each side requires
- This prevents `None` values from reaching `_custom_collate_fn` or model encode functions, without needing defensive guards throughout the codebase

Tested with `mteb run -m mteb/baseline-random-encoder -t CIRRIT2IRetrieval NIGHTSI2IRetrieval Fashion200kI2TRetrieval VisualNewsI2TRetrieval` — all complete successfully

### CODE REVIEW

**Comment by Samoed** [comment]
Maybe better to remove columns with None from tasks?

**Comment by Samoed** [comment]
Created pr for my suggestion https://github.com/embeddings-benchmark/mteb/pull/4446

File: mteb/models/model_implementations/random_baseline.py
Comment (KennethEnevoldsen) [inline mteb/models/model_implementations/random_baseline.py:1]: hmm wouldn't this resize the list and thus lead to an incorrect merge?

E.g. if we have the input:
```
input = {
text=["text1", "text2", "text-with-image"],
image=[None, None, Image3],
}
```

We would end up with an image3 getting merged with text1.

File: mteb/_create_dataloaders.py [hunk 1]
```diff
@@ -60,29 +60,28 @@ def _create_dataloader_from_texts(
 def _corpus_to_dict(
     row: dict[str, str],
 ) -> dict[str, str]:
-    text = (
-        (row["title"] + " " + row["text"]).strip()
-        if "title" in row and len(row["title"]) > 0
-        else row["text"].strip()
-    )
+    raw_text = row.get("text") or ""
```
Comment (KennethEnevoldsen) [inline mteb/_create_dataloaders.py]: not sure why this change is needed?

**Review by KennethEnevoldsen — COMMENTED** [review]
Hmm would really like a test on this behaviour

**Comment by Samoed** [comment]
I don't think that we want to handle such values, because we would need to update each model implementation and this would cause some corner cases, and we don't have such tasks for now

File: mteb/_create_dataloaders.py (hunk 1 above)
Comment (isaac-chung) [reply mteb/_create_dataloaders.py]: row['text'] can be None, so None.strip() will raise an error.

File: mteb/models/model_implementations/random_baseline.py
Comment (isaac-chung) [reply mteb/models/model_implementations/random_baseline.py:1]: True, will undo this.

**Comment by isaac-chung** [comment]
Reverted the random baseline encoder change.

**Comment by isaac-chung** [comment]
> I don't think that we want to handle such values, because we would need to update each model implementation and this would cause some corner cases, and we don't have such tasks for now

@Samoed I followed your advice, and updated the PR. Here is an example.

For VisualNewsI2TRetrieval (category i2t): query modalities = ["image"], doc modalities = ["text"].         
                                                                                                            
Queries (image-only):                                                                                       
  1. Dataset has columns: id, text, image (where text is None for image-only queries)
  2. _prepare_dataset called with prompt_type=PromptType.query                                                
  3. get_modalities(query) → ["image"]                                                                        
  4. "text" in modalities is False → skip _combine_queries_with_instruction_text                              
  5. "image" in modalities is True → _prepare_image_dataset runs, decodes images                              
  6. New code: drops columns not in ["image"] → removes text column (which had None values)
  7. Final dataset columns: id, image — no None values, collate works fine

Corpus (text-only):
  1. Dataset has columns: id, text, image (where image is None for text-only docs)
  2. _prepare_dataset called with prompt_type=PromptType.document
  3. get_modalities(document) → ["text"]
  4. "text" in modalities is True → _corpus_to_dict runs, producing id, text, body, optionally title
  5. "image" in modalities is False → skip _prepare_image_dataset
  6. New code: drops columns not in ["text"] → removes image column (which had None values)
  7. Final dataset columns: id, text, body — no None values, collate works fine

**Comment by Samoed** [comment]
We need to also handle title, instruction, query, conversation etc

**Comment by isaac-chung** [comment]
> We need to also handle title, instruction, query, conversation etc

That is is valid but not triggered by the 4 failing tasks in #4436. Could support in sep PR?

**Review by Samoed — APPROVED** [review]
Yes, I misunderstood a bit firstly. Looks good
